posix: avoiding redundant access of dictionary

rkothiya opened the pull request:
This patch fixes the redundant access of dictionary
for the same information by the macro PL_LOCAL_GET_REQUESTS

fixes: #1707

Change-Id: I48047537436ce920e74bc11cecd9773d7fe4457c
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
posix: avoiding redundant access of dictionary

This patch fixes the redundant access of dictionary
for the same information by the macro PL_LOCAL_GET_REQUESTS

fixes: #1707

Change-Id: I48047537436ce920e74bc11cecd9773d7fe4457c
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

rkothiya commented:
/run regression

mykaul commented on xlators/features/locks/src/posix.c:
```diff
@@ -172,7 +177,7 @@ fetch_pathinfo(xlator_t *, inode_t *, int32_t *, char **);
     })
 
 gf_boolean_t
```
```suggestion
static gf_boolean_t
```

mykaul commented on xlators/features/locks/src/posix.c:
```diff
@@ -208,23 +217,23 @@ dict_delete_domain_key(dict_t *dict, char *key, data_t *value, void *data)
 }
 
 void
```
```suggestion
static void
```

rkothiya commented:
/run regression

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -93,9 +93,14 @@ fetch_pathinfo(xlator_t *, inode_t *, int32_t *, char **);
 #define PL_LOCAL_GET_REQUESTS(frame, this, xdata, __fd, __loc, __newloc)       \
     do {                                                                       \
-        if (pl_has_xdata_requests(xdata)) {                                    \
+        unsigned char bitfield = 0;                                            \
```
I think it should be better to use an `unsigned int` ?

rkothiya replied:
I have used unsigned char because we only have 6 bits to store.

xhernandez replied:
Using a char doesn't provide any benefit. In fact processors tend to work better when using full 32 or 64 bits registers. If the reason for using a char is for optimization, then use a regular int, please.

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -193,11 +198,15 @@ pl_has_xdata_requests(dict_t *xdata)
+    if (*bitfield > 0)
+        return _gf_true;
+    else
+        return _gf_false;
```
I think it would be better to directly return `bitfield`. The caller can check if it's 0 or not, and we avoid an argument by reference, which requires memory storage. As a return value, it can be stored in registers.

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -207,24 +216,24 @@ dict_delete_domain_key(dict_t *dict, char *key, data_t *value, void *data)
     local->xdata = dict_copy_with_ref(xdata, NULL);
 
-    if (dict_get_sizen(xdata, GLUSTERFS_ENTRYLK_COUNT)) {
+    if (IS_BIT_SET(bitfield, 0)) {
```
It would be much better (and safe) to use macros or an enum instead of constants. And use the same macros/enums when initializing the `reqs` and `reqs_size` arrays in line 182 (like `[GLUSTERFS_ENTRYLK_COUNT_BIT] = GLUSTERFS_ENTRYLK_COUNT`)

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -207,24 +216,24 @@ dict_delete_domain_key(dict_t *dict, char *key, data_t *value, void *data)
-    if (dict_get_sizen(xdata, GLUSTERFS_ENTRYLK_COUNT)) {
+    if (IS_BIT_SET(bitfield, 0)) {
         local->entrylk_count_req = 1;
         dict_del_sizen(xdata, GLUSTERFS_ENTRYLK_COUNT);
```
We could improve this even more by doing a small change in `dict_del_*()` functions. We could return a boolean indicating whether the value was present or not. Using this new functionality, we can change the `dict_getn()` at line 202 by a `dict_deln()` and we won't need to do it here.

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -207,24 +216,24 @@ dict_delete_domain_key(dict_t *dict, char *key, data_t *value, void *data)
 
-    if (dict_get_sizen(xdata, GLUSTERFS_ENTRYLK_COUNT)) {
+    if (IS_BIT_SET(bitfield, 0)) {
         local->entrylk_count_req = 1;
```
Why we don't simply store the `bitfield` and we get rid of these booleans ?

rkothiya replied:
Can you please elaborate.

xhernandez replied:
Instead of keeping a boolean in `local` for each value we need, we can store the bitfield itself in a single field and then check the corresponding bit instead of the current boolean field.

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -189,15 +194,30 @@ pl_has_xdata_requests(dict_t *xdata)
+    if (dict_del_sizen(xdata, GLUSTERFS_MULTIPLE_DOM_LK_CNT_REQUESTS))
+        SET_BIT(bitfield, GLUSTERFS_MULTIPLE_DOM_LK_CNT_REQUESTS_BIT);
+
+    if (dict_del_sizen(xdata, GLUSTERFS_INODELK_DOM_COUNT))
+        SET_BIT(bitfield, GLUSTERFS_INODELK_DOM_COUNT_BIT);
```
We cannot delete this one here. This key contains a string that we need to keep. In this particular case we need to do a `dict_get_sizen()`

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -93,9 +93,14 @@ fetch_pathinfo(xlator_t *, inode_t *, int32_t *, char **);
 
+#define IS_BIT_SET(byte, bit_number) (byte & (1 << bit_number))
+
+#define SET_BIT(byte, bit_number) (byte = (byte | (1 << bit_number)))
```
Modifying arguments inside a macro is a bad practice because it's a hidden modification and because `byte` could be anything, even thing that cannot appear in the left side of an assignment.

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -189,15 +194,31 @@ pl_has_xdata_requests(dict_t *xdata)
+    if (dict_del_sizen(xdata, GLUSTERFS_MULTIPLE_DOM_LK_CNT_REQUESTS))
+        SET_BIT(bitfield, GLUSTERFS_MULTIPLE_DOM_LK_CNT_REQUESTS_BIT);
+
+    if (dict_del_sizen(xdata, GLUSTERFS_INODELK_DOM_COUNT))
```
This should be `dict_getn`

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -189,15 +194,31 @@ pl_has_xdata_requests(dict_t *xdata)
+    if (dict_del_sizen(xdata, GLUSTERFS_INODELK_DOM_COUNT))
+        SET_BIT(bitfield, GLUSTERFS_INODELK_DOM_COUNT_BIT);
+
+    if (dict_getn(xdata, GLUSTERFS_POSIXLK_COUNT,
```
This should be `dict_del_sizen`

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -255,9 +255,7 @@ pl_needs_xdata_response(pl_local_t *local)
-    if (local->parent_entrylk_req || local->entrylk_count_req ||
-        local->inodelk_dom_count_req || local->inodelk_count_req ||
-        local->posixlk_count_req || local->multiple_dom_lk_requests)
+    if (local->bitfield)
```
Where is ` local->bitfield` assigned ?

xhernandez reviewed (changes requested):
When you update the patch, please don't amend the previous commit. Create a new commit. It's much easier and faster to review.

Commit:
posix: avoiding redundant access of dictionary

- Converted the macro SET_BIT to function set_bit
- Removed the code to delete the key GLUSTERFS_INODELK_DOM_COUNT
- Assigned the value to local->bitfield

Change-Id: I101f3fda65e9e75e05907d671203c5d7f072fa8f
Fixes: #1707
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

xhernandez commented on xlators/features/locks/src/posix.c:
```diff
@@ -235,17 +239,6 @@ pl_get_xdata_requests(pl_local_t *local, dict_t *xdata)
                                                   GLUSTERFS_INODELK_DOM_COUNT);
     if (local->inodelk_dom_count_req) {
         data_ref(local->inodelk_dom_count_req);
-        dict_del_sizen(xdata, GLUSTERFS_INODELK_DOM_COUNT);
```
The key needs to be deleted if present.

Commit:
posix: avoiding redundant access of dictionary

deleted GLUSTERFS_INODELK_DOM_COUNT key

Change-Id: I638269e6a9f6fc11351eaede4c103e032881fe12
Fixes: #1707
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

rkothiya commented:
/run regression

mohit84 commented:
/recheck smoke

mohit84 commented:
@rkothiya 
/home/jenkins/root/workspace/gh_centos7-regression/libglusterfs/src/dict.c: In function 'dict_del':
15:41:58 /home/jenkins/root/workspace/gh_centos7-regression/libglusterfs/src/dict.c:613:5: error: 'return' with a value, in function returning void [-Werror]
15:41:58      return dict_deln(this, key, strlen(key));
15:41:58      ^
15:41:58 /home/jenkins/root/workspace/gh_centos7-regression/libglusterfs/src/dict.c: In function 'dict_deln':
15:41:58 /home/jenkins/root/workspace/gh_centos7-regression/libglusterfs/src/dict.c:626:9: error: 'return' with no value, in function returning non-void [-Werror=return-type]
15:41:58          return;
15:41:58          ^
15:41:58 cc1: all warnings being treated as errors
15:41:58 make[3]: *** [libglusterfs_la-dict.lo] Error 1
15:41:58 make[3]: *** Waiting for unfinished jobs....
15:41:58 make[3]: Leaving directory `/d/build/scratch/libglusterfs/src'
15:41:58 make[2]: *** [install] Error 2
15:41:58 make[2]: Leaving directory `/d/build/scratch/libglusterfs/src'
15:41:58 make[1]: *** [install-recursive] Error 1
15:41:58 make[1]: Leaving directory `/d/build/scratch/libglusterfs'

regression is failing due to function void, kindly resolve it.

mykaul commented:
I believe he's on PTO until the 18th, so it'll take some time to be resolved.

Commit:
posix: avoiding redundant access of dictionary

Smoke test warnings fixed.

Fixes: #1707
Change-Id: I8682bd0e49f44cbc1442324e1756b56481f18ccd
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
Done

rkothiya commented:
/recheck smoke

rkothiya commented:
/recheck smoke

rkothiya commented:
/run regression

deepshikhaaa commented:
/recheck smoke
